pkg/storage/gcs/gcp_labels_tags.go

```diff
@@ -35,3 +77,213 @@ func getUserLabels(infraLister configlisters.InfrastructureLister) (map[string]s
+	limiter := rate.NewLimiter(rate.Every(time.Second/time.Duration(limit)), burst)
+
+	if emptyBucket {
+		limiter.AllowN(time.Now(), burst)
+	}
+
+	return limiter
+}
+
+// toTagValueList converts the tags to an array containing tagValues
+// NamespacedNames.
+func toTagValueList(tags []configv1.GCPResourceTag) []string {
+	if len(tags) <= 0 {
+		return nil
+	}
+
+	list := make([]string, 0, len(tags))
+	for _, tag := range tags {
+		t := fmt.Sprintf("%s/%s/%s", tag.ParentID, tag.Key, tag.Value)
+		list = append(list, t)
+	}
+	return list
+}
+
+// getInfraResourceTagsList returns the user-defined tags present in the
+// status sub-resource of Infrastructure.
+func getInfraResourceTagsList(platformStatus *configv1.PlatformStatus) []configv1.GCPResourceTag {
+	if platformStatus != nil && platformStatus.GCP != nil && platformStatus.GCP.ResourceTags != nil {
+		return platformStatus.GCP.ResourceTags
+	}
+	klog.V(1).Infof("getInfraResourceTagsList: user-defined tag list is not provided")
+	return nil
+}
+
+// getTagsList returns the list of tags to apply on the resources.
+func getTagsList(platformStatus *configv1.PlatformStatus) []string {
+	return toTagValueList(getInfraResourceTagsList(platformStatus))
+}
+
+// getFilteredTagList returns the list of tags to apply on the resources after
+// filtering the tags already existing on a given resource.
+func getFilteredTagList(ctx context.Context, platformStatus *configv1.PlatformStatus, client *rscmgr.TagBindingsClient, parent string) []string {
+	return filterTagList(ctx, client, parent, getTagsList(platformStatus))
+}
+
+// filterTagList returns the filtered list of tags to apply on the resources.
+func filterTagList(ctx context.Context, client *rscmgr.TagBindingsClient, parent string, tagList []string) []string {
+	dupTags := make(map[string]bool, len(tagList))
+	for _, k := range tagList {
+		dupTags[k] = false
+	}
+
+	listBindingsReq := &rscmgrpb.ListEffectiveTagsRequest{
+		Parent: parent,
+	}
+	bindings := client.ListEffectiveTags(ctx, listBindingsReq)
+	// a resource can have a maximum of {gcpMaxTagsPerResource} tags attached to it.
+	// Will iterate for {gcpMaxTagsPerResource} times in the worst case scenario, if
+	// none of the break conditions are met. Should the {gcpMaxTagsPerResource} be
+	// increased in future, it should not create an issue, since this is an optimization
+	// attempt to reduce the number the tag write calls by skipping already existing tags,
+	// since it has a quota restriction.
+	for i := 0; i < gcpMaxTagsPerResource; i++ {
+		binding, err := bindings.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil || binding == nil {
+			klog.V(1).Infof("failed to list effective tags on the %s bucket: %v: %v", parent, binding, err)
+			break
+		}
+		tag := binding.GetNamespacedTagValue()
+		if _, exist := dupTags[tag]; exist {
+			dupTags[tag] = true
+			klog.V(1).Infof("filterTagList: skipping tag %s already exists on the %s bucket", tag, parent)
```

Comment:
Same here, this looks like a debug log to me, so consider using `V(4)`

Reply:
suggestion incorporated.